### ISSUE TITLE
Unable to get/create model instance from data in the store

### ISSUE THREAD

**Issue by morphatic** [feathersjs-ecosystem/feathers-vuex#265]
This is weird. I'm working on a form to edit the data in a "Conference" (think sports league conferences). I know that the conferences are all loaded into the store because I've got a "Conferences" page with a data table that shows all of them. The problem happens when I click on a conference to go to the page where I can view/edit its details.

In order to be able to edit it, I followed the normal practice of cloning an instance, modifying the clone, and then calling `save()` on the clone. However, I'm unable to retrieve the conference as an instance of the model class, and therefore I'm getting the error "`clone()` is undefined" when I get to that step. Here's roughly what I've got:

```js
created () {
  const { Conference } = this.$FeathersVuex.api
  const original = Conference.getFromStore(this.$route.params.id)
  this.conference = original.clone()
}
```

I expected that since I'm using the model class to retrieve the conference from the store that the returned object would be an instance of `Conference`, but `original instanceof Conference` returns `false`. No worries, all the properties are there, so I'll just instantiate a new one, like this:

```js
created () {
  const { Conference } = this.$FeathersVuex.api
  const original = Conference.getFromStore(this.$route.params.id)
  if (!(original instanceof Conference)) {
    original = new Conference(original)
  }
  console.log(original instanceof Conference) // <-- returns FALSE!!! 😒
  this.conference = original.clone() // <-- therefore clone() is still undefined
}
```

Grrr. Okay maybe, I thought, it was because all of the reactive properties that Vue attached to `original`, so let's strip those out before instantiating the new conference object:

```js
created () {
  const { Conference } = this.$FeathersVuex.api
  const original = Conference.getFromStore(this.$route.params.id)
  if (!(original instanceof Conference)) {
    original = JSON.parse(JSON.stringify(original))
    console.log(original) // <-- yep, reactive props have been stripped away
    original = new Conference(original)
  }
  console.log(original instanceof Conference) // <-- still returns FALSE!!! 😒
  this.conference = original.clone() // <-- therefore clone() is still undefined
}
```

To be sure, I confirmed that I can create a new **empty** Conference that has all of the correct properties I set up in `instanceDefaults()`:

```js
const { Conference } = this.$FeathersVuex.api
const newConf = new Conference()
console.log(newConf instanceof Conference) // <-- TRUE 😄
```

Here's how the model is defined on the API side (using mongoose):

```js
const ConferenceSchema = new Schema({
  name: { type: String, required: true },
  aliases: [String],
  slug: { type: String, required: true, index: true, unique: true },
  abbr: { type: String, required: true },
  colors: {
    primary:   { type: String, default: '#000000' },
    secondary: { type: String, default: '#ffffff' },
    tertiary:  { type: String, default: '#777777' },
    other:     { type: String, default: '' }
  },
  logo: String,
  website: { type: String, validator: v => is_valid_url.test(v) },
  teams: [{ type: Schema.Types.ObjectId, ref: 'Team' }]
}, {
  timestamps: true
})
```

And here is how the model is defined for `feathers-vuex` on the client side:

```js
import { api, makeServicePlugin, BaseModel } from '@/services/api'

class Conference extends BaseModel {
  static modelName = 'Conference'
  static instanceDefaults = () => ({
    name: '',
    aliases: null,
    slug: '',
    abbr: '',
    colors: {
      primary: '#000000',
      secondary: '#ffffff',
      tertiary: '#777777',
      other: ''
    },
    logo: null,
    website: null,
    teams: null,
    createdAt: null,
    updatedAt: null
  })
}

const servicePath = 'conferences'
const servicePlugin = makeServicePlugin({
  Model: Conference,
  service: api.service(servicePath),
  servicePath
})

export default servicePlugin
```

Am I missing something? Does this maybe have something to do with the array of `Team` ids?

### System configuration

**Module versions**:

* `@vue/cli`: ^3.9.0
* `vue`: ^2.6.10
* `vuex`: ^3.0.1
* `@feathersjs`: ^4.3.2
* `feathers-vuex`: 2.0.0-pre.77

**NodeJS version**: 10.15.3

**Operating System**: OSX 10.13.6 (High Sierra)

**Browser Version**: Chrome 77.0.3865.75

**Module Loader**: webpack (whatever `@vue/cli` uses)

**Comment by morphatic** [comment]
Okay, so I dug into the `feathers-vuex` source a bit, and I have some more insight, and maybe a solution. First, the potential solution looks like:

```js
created () {
  const { Conference } = this.$FeathersVuex.api
  const original = Conference.getFromStore(this.$route.params.id)
  this.conference = new Conference(original, { clone: true }) 
}
```

I believe the source of the problem is that the items in my store are POJOs and not `Conference` instances. This is because I keep most of the sports team data in an IndexedDB that gets loaded into the store whenever the user logs in or the browser window is refreshed or loaded for the first time. IndexedDB can only store objects, not class instances, so I lose the `feathers-vuex` functionality when I store them there.

Before, I had been re-hydrating the objects as model instances one-by-one as they got added back into the store. But since there are 1000s of objects in the store that was killing the responsiveness of the site. I was crashing the browser with out-of-memory errors.

So why didn't `new Conference(original)` work? It turns out, that in the event the `BaseModel.constructor()` is being used to create an instance that's already in the store, it [pulls the existing item out of the store](https://github.com/feathers-plus/feathers-vuex/blob/typescript/src/service-module/make-model.ts#L109) instead of going through the work of instantiating an entirely new object. Since in my case, the item in the store was a POJO, that's what I got. If `original` had been an entirely new object that was **not** already in the store, `new Conference(original)` would return an instance of `Conference` as expected.

As it turns out, I like the solution better than my original (buggy) attempt. All of this has given me a couple of ideas.

### Proposal 1: `cloneFromStore(id)`

As I understand it, the normal order of operations on "edit" pages is:

1. Retrieve an instance of a model
2. Clone it
3. Edit and save the clone

This means creating an intermediate variable to hold the item to be cloned, e.g. the `original` variable, that never gets used again, like this:

```js
created () {
  const { Conference } = this.$FeathersVuex.api
  const original = Conference.getFromStore(this.$route.params.id)
  this.conference = original.clone()
}
```

What if instead this were:

```js
created () {
  const { Conference } = this.$FeathersVuex.api
  this.conference = Conference.cloneFromStore(this.$route.params.id)
}
```

Maybe this is API bloat. Conceptually, it took me a while to get used to the retrieve-clone-edit-save pattern. Perhaps this would be easier for newcomers? I think the practice of only ever editing clones is really smart. This might make it clearer. Or confuse people more. I don't feel too strongly about it. It's mostly just an idea to consider.

### Proposal 2: Add type-checking to `BaseModel.constructor()`

The use case I described was probably not part of `feathers-vuex` design. Although it's on somewhat of a hiatus, I'm really enamored with the ["offline first" approach](https://auk.docs.feathersjs.com/guides/offline-first/readme.html) that was a part of Auk. LocalForage (IndexedDB) makes it pretty easy to replicate large parts of your database locally, but since you can't store class instances (only objects), there has to be some mechanism to rehydrate them when they actually get used in an app.

So in cases where the constructor gets called with an existing item that is just a POJO, we might could do something like:

```js
// service-module/make-model.ts (around line 113 at time of writing)
// If it already exists, update the original and return
if (existingItem) {
  // if the existing item is NOT a model instance
  if (!(existingItem instanceof BaseModel)) {
    // convert it into one
    existingItem = Object.assign(this, existingItem)
  }
  data = setupInstance.call(this, data, { models, store }) || data
  _commit.call(this.constructor, 'mergeInstance', data)
  return existingItem
}
```

Caveat: I haven't tested this code, so don't know if it would work as-is. I also don't know if there are other places we'd need to make sure the return type is a model instance. I think this is more of a longer-term idea. And it may be a bad one. Again, I'm just throwing it out there.

Anyway, I think I've solved my immediate problem. I'll leave this open in case @marshallswain wants to comment on the proposals. Otherwise, it's fine to close.

**Comment by marshallswain** [comment]
@morphatic the solution that I would like to see in place is along the lines of your type-checking proposal.  Your solution would probably work AFTER using a mutation to replace the record in the store with the new instance version of the same data.

**Comment by marshallswain** [comment]
I've put a solution to this in https://github.com/feathersjs-ecosystem/feathers-vuex/issues/386.  That solution is a front-loaded upgrade of all data in the store into instances.  I can't imagine it being very resource intensive to upgrade a few records, so once we implement it we might be able to close this issue.  If there's a use case where we need to hydrate thousands of records, then we might need to implement the incremental solution.

**Comment by marshallswain** [comment]
Fixed in #401 

Fixed in `feathers-vuex@3.3.0`
